### PR TITLE
feat: inference feats and simulator

### PR DESCRIPTION
**Pull request by pixix**
Fixes #

### CODE REVIEW

File: src/lambda.d/simulator/gen.py [hunk 1]
```diff
@@ -11,13 +12,28 @@
 logger.setLevel(logging.INFO)
 
 sqs = boto3.client('sqs')
```
Comment (zxkane) [inline src/lambda.d/simulator/gen.py:14]: remove unused sqs code

File: src/lib/dashboard-stack.ts [hunk 2]
```diff
@@ -334,13 +336,24 @@ export class TransactionDashboardStack extends NestedStack {
...
         QUEUE_URL: props.queue.queueUrl,
         DATASET_URL: getDatasetMapping(this).findInMap(Aws.PARTITION, IEEE),
       },
       timeout: Duration.minutes(15),
       memorySize: 3008,
     });
     props.queue.grantSendMessages(tranSimFn);
+    tranSimFn.addToRolePolicy(new PolicyStatement({
+      actions: ['lambda:InvokeFunction'],
+      resources: [
```
Comment (zxkane) [inline src/lib/dashboard-stack.ts:349]: why not using existing `inferenceArn`?

File: src/lib/dashboard-stack.ts [hunk 3]
```diff
@@ -334,13 +336,24 @@ export class TransactionDashboardStack extends NestedStack {
...
       runtime: Runtime.PYTHON_3_8,
       environment: {
+        INFERENCE_ARN: String(props.inferenceArn),
         QUEUE_URL: props.queue.queueUrl,
         DATASET_URL: getDatasetMapping(this).findInMap(Aws.PARTITION, IEEE),
       },
       timeout: Duration.minutes(15),
       memorySize: 3008,
     });
     props.queue.grantSendMessages(tranSimFn);
```
Comment (zxkane) [inline src/lib/dashboard-stack.ts:346]: remove this useless permission

File: src/lib/inference-stack.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      memorySize: 512,
+      environment: {
+        MAX_FEATURE_NODE: String(10),
+        CLUSTER_ENDPOINT: props.neptune.endpoint,
+        CLUSTER_PORT: props.neptune.port,
+        CLUSTER_REGION: String(this.region),
+        ENDPOINT_NAME: endpointName,
+        MODEL_BTW: String(0.5),
+        QUEUE_URL: props.queue.queueUrl,
+        TRANSACTION_ID_COLS: 'card1,card2,card3,card4,card5,card6,ProductCD,addr1,addr2,P_emaildomain,R_emaildomain',
```
Comment (zxkane) [inline src/lib/inference-stack.ts:53]: should share the column lists with training stack,

https://github.com/aws-samples/realtime-fraud-detection-with-gnn-on-dgl/blob/main/src/lib/etl-glue.ts#L190-L191

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/lambda.d/inference/func/inferenceApi.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+GremlinUtils.init_statics(globals())
+endpoints = Endpoints(neptune_endpoint = CLUSTER_ENDPOINT, neptune_port = CLUSTER_PORT, region_name = CLUSTER_REGION)
+gremlin_utils = GremlinUtils(endpoints)
+conn = gremlin_utils.remote_connection()
+g = gremlin_utils.traversal_source(connection=conn)
+
+def load_data_from_event(input_event, transactions_id_cols, identities_cols, neighbor_cols, dummied_col):
+    """Load and transform event data into correct format for next step subgraph loading and model inference input. 
+        input event keys should come from related dataset and following the requirement 
+        from ['https://quip-amazon.com/gWm8AydvuN5O/Data-Schema-Details']
```
[the remark this answers is not in the document]
Comment (pixix) [reply src/lambda.d/inference/func/inferenceApi.py:40]: ok

File: src/lib/dashboard-stack.ts (hunk 3 above)
Comment (pixix) [reply src/lib/dashboard-stack.ts:346]: it is the FunctionName for the boto3 inside to invoke lambda?

Comment (pixix) [reply src/lib/dashboard-stack.ts:346]: removed useless sqs permission

File: src/lambda.d/simulator/gen.py [hunk 6]
```diff
@@ -39,37 +53,22 @@ def handler(event, context):
         
         sampleDF = mergedDF.sample()
         
-        sampleData = sampleDF.iloc[0]
+        sampleDF.columns = [x.replace('-','_') if '-' in x else x for x in sampleDF.columns]
```
Comment (zxkane) [inline src/lambda.d/simulator/gen.py:56]: why do this transforming?

File: src/lambda.d/inference/func/inferenceApi.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    logger.info(f'Send transaction {data_output} to queue.')
+    response = sqs.send_message(
+        QueueUrl=QUEUE_URL,
+        DelaySeconds=0,
+        MessageBody=json.dumps(data_output),
+        MessageGroupId=context.aws_request_id,
+    )
+    
+    return f"Done. with flag:{pred_prob> MODEL_BTW} and pred_prob:  {pred_prob}"
```
Comment (zxkane) [inline src/lambda.d/inference/func/inferenceApi.py:301]: use json object as response

File: src/lib/etl-glue.ts [hunk 8]
```diff
@@ -187,7 +189,7 @@ export class ETLByGlue extends Construct {
         '--database': transactionDatabase.databaseName,
         '--transaction_table': transactionTable.tableName,
         '--identity_table': identityTable.tableName,
-        '--id_cols': 'card1,card2,card3,card4,card5,card6,ProductCD,addr1,addr2,P_emaildomain,R_emaildomain',
+        '--id_cols': this.preprocessingJob_id_cols,
         '--cat_cols': 'M1,M2,M3,M4,M5,M6,M7,M8,M9',
```
Comment (zxkane) [inline src/lib/etl-glue.ts:193]: also extract this variable to parent stack. it is probably used by inference method

File: src/lib/inference-stack.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      memorySize: 512,
+      environment: {
+        MAX_FEATURE_NODE: String(10),
+        CLUSTER_ENDPOINT: props.neptune.endpoint,
+        CLUSTER_PORT: props.neptune.port,
+        CLUSTER_REGION: String(this.region),
+        ENDPOINT_NAME: endpointName,
+        MODEL_BTW: String(0.9),
+        QUEUE_URL: props.queue.queueUrl,
+        TRANSACTION_ID_COLS: String(props.preprocessingJob_id_cols),
```
Comment (zxkane) [inline src/lib/inference-stack.ts:54]: why add `String()` again

File: src/lib/inference-stack.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      index: 'inferenceApi.py',
+      runtime: Runtime.PYTHON_3_8,
+      handler: 'handler',
+      timeout: Duration.minutes(2),
+      memorySize: 512,
+      environment: {
+        MAX_FEATURE_NODE: String(10),
+        CLUSTER_ENDPOINT: props.neptune.endpoint,
+        CLUSTER_PORT: props.neptune.port,
+        CLUSTER_REGION: String(this.region),
```
Comment (zxkane) [inline src/lib/inference-stack.ts:50]: not sure if `this.region` referring the concrete region, use `Aws.REGION` instead

File: src/lib/etl-glue.ts [hunk 11]
```diff
@@ -176,6 +177,7 @@ export class ETLByGlue extends Construct {
       },
     });
     const outputPrefix = `${props.s3Prefix ?? ''}processed-data/`;
+    this.preprocessingJob_id_cols = 'card1,card2,card3,card4,card5,card6,ProductCD,addr1,addr2,P_emaildomain,R_emaildomain';
```
Comment (zxkane) [inline src/lib/etl-glue.ts:180]: I prefer to move this constant to parent stack, also rename it to `datasetIdCols`

File: test/dashboard-stack.test.ts [hunk 12]
```diff
@@ -1085,10 +1083,12 @@ function initializeStackWithContextsAndEnvs(context: {} | undefined, env?: {} |
     removalPolicy: RemovalPolicy.DESTROY,
     visibilityTimeout: Duration.seconds(60),
   });
+  const inferenceArn = new String();
```
Comment (zxkane) [inline test/dashboard-stack.test.ts:1086]: use a string like `arn:aws:lambda:ap-southeast-1:123456789012:function:inference-func`

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/lib/dashboard-stack.ts [hunk 13]
```diff
@@ -343,13 +346,17 @@ export class TransactionDashboardStack extends NestedStack {
       index: 'gen.py',
       runtime: Runtime.PYTHON_3_8,
       environment: {
-        QUEUE_URL: props.queue.queueUrl,
+        INFERENCE_ARN: inferenceStatsFnArn,
```
Comment (zxkane) [inline src/lib/dashboard-stack.ts:349]: use `props.inferenceArn` directly

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/lib/inference-stack.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+export class InferenceStack extends NestedStack {
+  readonly inferenceSG: ISecurityGroup;
+  readonly inferenceStatsFn: IFunction;
+
+  constructor(scope: Construct, id: string, props: InferenceProps) {
+    super(scope, id, props);
+
+    this.inferenceStatsFn = new PythonFunction(this, 'InferenceStatsFn', {
```
Comment (zxkane) [inline src/lib/inference-stack.ts:32]: rename `InferenceStatsFn` to `InferenceFn`

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/lib/inference-stack.ts (hunk 14 above)
Comment (zxkane) [inline src/lib/inference-stack.ts:32]: enable trace for lambda func

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**